### DISCUSSION TITLE
Resize Canvas and zoom factor according to ViewPort current size

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#398] [Q&A] [answered]
Hi everybody, Merry Christmas!

I updated some source files of Ctrlr that let the panel ViewPort (the window of the panel in the host) to be resized (or not) with the Juce bottom-right resizer corner.

[CtrlrEditor.cpp](https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/UIComponents/CtrlrApplicationWindow/CtrlrEditor.cpp) @ Line #40

```
	if (!JUCEApplication::isStandaloneApp()) 
	{ 
	    if (owner.getInstanceMode() == InstanceSingleRestriced) 
        { 
        	/* Restricted instances don't get to be resizable */ 
        	setResizable(false, false);  // ViewPort Locked, must be set to (true, false) to make it resizable
        } 
        else 
        { 
	    addAndMakeVisible (&resizer); 
            resizer.setAlwaysOnTop (false); 
            resizer.grabKeyboardFocus(); 
            resizer.toFront (true); 
        } 
	}
```

In Ctrlr the initial ViewPort size is defined by the Global Canvas bounds (top pos, left pos, width, height) of our Panels. 
We can apply a zoom factor to increase/decrease the size of the Canvas and its child elements.

Let's say the Panel has an aspect ratio of 4:3   with a standard size at zoom = 1.0 for 800x600px

How to change the zoom ratio relative to the ViewPort size (with the resizer corner enabled), while keeping the aspect ratio?

I thought about this feature for a long time to get exported panels like the one from Softube or fabfilter. They are responsive, when you drag the corner the panel zooms to fit the ViewPort while keeping the defined aspect ratio.

i check the Juce documentation and Forum, there's plenty of posts about that.
But in Ctrlr I'm not sure about the relationship between the Canvas Size and the ViewPort size. 
Or if even it's possible to get the current ViewPort Size with LUA.
if we could get the ViewPort size with LUA we could create a relation between the zoom factor of the Canvas and the instance window (viewport), that would be so nice!

Thanks for your insight about that.

Have a wonderful Christmas time

Damien

**Comment by Godlike-Productions** [comment]
I'd look at something like

```
origWidth = panel:getCanvas():getWidth()
origHeight = panel:getCanvas():getHeight()
windowWidth = panel:getCanvas():getParentWidth()
windowHeight = panel:getCanvas():getParentHeight() -- or panel:getPanelEditor():getProperty("uiPanelViewPortSize")

widthZoom = windowWidth/origWidth
heightZoom = windowHeight/origHeight

zoomFactor = math.min(widthZoom, heightZoom)

panel:getPanelEditor():setProperty("uiPanelZoom",zoomFactor,false) --basically set zoom to the smallest zoom factor to make sure it always fits in the window.


```

I might give this a go on the next update of the Jupiter 8 Plugin.

You would need to have a callback for onResize and I'm not sure how to do that. It's not listed in the panel properties in 5.3.201

Also, this may be of use (though I don't know how to use it)

>>> panel:getCanvas():setBoundsToFit()
ERROR: No matching overload found, candidates:
void setBoundsToFit(Component&,int,int,int,int,Justification,bool)

From what I've seen of Juce, this is handled by either building stuff in vectors or by setting the size and position of each element relative to the window size

**Reply by damiensellier** [reply]
Hi jason, yes it seems that's all the parameter involved. Thanks a lot!
I'll try on my panel later on this week to see how it resizes and how to handle the save/restore process of the parameters with saveState()
I'm not sure if it will need some extra instructions from Juce added to Ctrlr itself for calling the refresh on resize of the top parent element containing the panel or the viewPort.

**Reply by Godlike-Productions** [reply]
Just checked. In 5.6.23 there is a luaPanelResized callback. You should be able to put this method there.

**Reply by damiensellier** [reply]
I tried a little bit and we probably need to implement an extra callback for when the viewPort changes.
luaPanelResized works really well when you are in editor mode and can resize the panel canvas by dragging the bottom right corner, (it passes the uiPanelViewPortSize for exemple) but what I need is the other way around.
We need a callback "luaViewPortResized" this way, the width of the application or plugin frame can send its value to the child (the panel canvas) and resize/zoom itself accordingly with the panel:getCanvas():repaint().

I'll check how the luaPanelResized callback works precisely in the Ctrlr Src. It can probably adapted to the viewPort settings.

Worst case scenario, the available functions can be used to resize the plugin container on restart if we apply a zoom factor from a settings menu and store it as a saveState string, then on relaunch we can grab those values in the trees pass them as a MB and if there's such values it can be used to define the viewport bounds.
It's going to help if the window is locked in restricted instance without the resizer. it's useless if you have the resizer enabled anyway.
I'm working on this to avoid having tiny little panels with giant HiDPI screens like 4k and more. This way we would be able just to resize the panel size by dragging the corner and zoom accordingly. Instead of having 1 version of a locked plugin for HD and another one for HiDPI/Retina screens.

EDIT: I thought about another way to update the canvas size to the uiPanelViewPortSize
We can store the uiPanelViewPortSize as a memoryblock
Run a timer when the panel is loaded ( that runs like a pseudo callback). 
This timer compares the current uiPanelViewPortSize in the global properties to the one stored at t-1 in the memoryblock (1 timer period before).
If it's different, the canvas has to adjust to the "new" uiPanelViewPortSize and update the memoryblock. and refresh.
If it runs every second or so I don't know if the load on the cpu will be high I'll check it later.

**Reply by Godlike-Productions** [reply]
I just tried it, and it kind of works (need to sort a few bugs for sizing out), but the luaPanelResized is where you drag the bottom corner of the panel in ctrlr, not the window itself

**Reply by damiensellier** [reply]
@Godlike-Productions Yes i got it working too, I'm checking right now how to make it work from the whole app window since the number is updated realtime in the uiPanelViewPortSize property and the uiPanelPropertiesSize as well.
uiPanelViewPortSize + uiPanelPropertiesSize + 10(separator column) = entire app window width

Edit : Ok I checked everything but it doesn't work this way, the size of uiPanelPropertiesSize and uiPanelViewPortSize are working together but are relative to the overall app window size with a common multiplier, so both are always changing while resizing the big window. This is why the properties panel get narrower if you resize the app window.
I'll need to pass the app window size (from the owner getWidth and getHeight in the Ctrlr source files) and create 2 new properties in the xml for "uiPanelEditor" so that we can get access to those.
Also I need to add a callback function for when the "owner" is resized.

This is the only way to achieve GUi responsive plugin instances.

**Reply by Godlike-Productions** [reply]
If you solve this, please post.  I don't know how to build the callback that is required.

**Reply by damiensellier** [reply]
Hi,
Since there's no way to make this work without a callback on the whole app window resize, I worked on the possibility to select the size of the plugin in a popup menu.
I struggled with the sequence order with panelLoaded before readState etc but I finally got it working.
I'll post a sample panel to show the process for others. This way you can select the size of the plugin panel, the size is saved in the saveState tree, and when you re-open the session or the app the size is changed accordingly. So there's no need to export 2 plugin with different size for HD or Retina.
It's all good! I'll share the panel later.

**Comment by damiensellier** [comment, marked as answer]
With the stock Ctrlr application, it's impossible to have responsive GUI plugins since the callback from the APP or plugin viewport resizer is not implemented.

**WORKAROUND**

Here is a Demo Panel with a popup triggered when the Label is clicked.
This popup shows how to select/save the midi inputs and outputs as well as a preset panel size.

1. Select your desired panel size.
2. close Ctrlr (not the panel) 
3. open Ctrlr and you'll see the effect. 

Since the panel in Ctrlr or as a standalone app can be resized without any workaround with the zoom and the bottom-right corner resizer, this demo mostly concerns the panel as exported restricted plugin (VST, VST3 or AU) version thanks to the saveState, readState scripts that stores parameters in the host session file.

[Select Window Size_1_0_0_2022-01-05_21-15.panel.zip](https://github.com/RomanKubiak/ctrlr/files/7819295/Select.Window.Size_1_0_0_2022-01-05_21-15.panel.zip)



```
panel:getCanvas():setSize(width,height)    -- has to be maintained. Whatever the current zoom is, keep it like the original size for zoom = 1.0
panel:setProperty("ctrlrEditorBounds","0 0 width*ZoomFactor height*ZoomFactor",true)  -- has to define the desired final vst window size (works within host)
panel:getPanelEditor():setProperty("uiPanelCanvasRectangle","0 0 width*ZoomFactor height*ZoomFactor",true) -- has to change accordingly (works within Ctrlr)
panel:getPanelEditor():setProperty("uiPanelZoom",ZoomFactor,true)  -- change the zoom factor
```

The WindowSizeValue will be set in the saveState Tree in Ctrlr.settings on quit or your DAW session file on save.
It looks like this when saved:
`<panelCustomData SaveStateDisplaySizeIndex="1"/>`
